### ISSUE TITLE
Custom Font Face

### ISSUE THREAD

**Issue by aqilhex** [PHPOffice/PhpSpreadsheet#2760]
Does PhpSpreadsheet Support custom font face?I have an installed custom font face on my windows machine. Whenever i try to modify an excel file and export it, The font face won't apply on cell text.

**Comment by MarkBaker** [comment]
Custom Font Face? What's that? It's not something I'm familiar with outside of CSS.
Excel allows you to set the font that you use for text in a cell; based on the fonts available on the machine where Excel is running; but as far as I'm aware it isn't possible to embed fonts in an Excel document in the way you can with (for example) a PDF

**Comment by aqilhex** [comment]
You are misunderstanding me. I have a template.xlsx, Whenever i try to modify and save it, Which i should use write method and create new file, The format of template.xlsx is differ from the new one created.

**Comment by MarkBaker** [comment]
I am misunderstanding: What format are you saving? What's different? My guessing isn't very good this morning

**Comment by aqilhex** [comment]
Let me explain, I have this template.xlsx with some custom style. I have installed custom font face on my windows machine which name is "B Nazanin". Clear? Now when i try to modify template.xlsx which below code:
`
$reader = new \PhpOffice\PhpSpreadsheet\Reader\Xlsx();
$spreadsheet = $reader->load("./excel/template.xlsx");

$spreadsheet->getActiveSheet()->setCellValue("H11","Blah Blah");

$writer = \PhpOffice\PhpSpreadsheet\IOFactory::createWriter($spreadsheet, 'Xlsx');
$writer->save('./excel/2.xlsx');
`
I get the new file exported at the same directory and the data successfully inserted and everything is fine. But the style is broken in the exported file (2.xlsx).

I tried to replace template.xlsx but the issue persist.

**Comment by MarkBaker** [comment]
In what way is the style broken? How does that font appear when you display it?

**Comment by aqilhex** [comment]
The style of template.xlsx not get applied on the exported one. Which mean, the exported file 2.xlsx displaying me "Calibri" font name and not B Nazanin which i applied on template file.

**Comment by MarkBaker** [comment]
I've just tested. Created a spreadsheet and set a coupe of cells to use a font I have installed that isn't part of a standard Windows distribution; loaded that spreadsheet with PhpSpreadsheet and saved it with a new name. When I open the new file, the cells are still displayed in the correct font. So it works as expected.

MS Excel stores the name of the font in its styles; and when you display that spreadsheet on a different machine that may not have the font installed, it will try to find a similar font installed to use instead, and it will only fall back to using Calibri/Ariel if it can't find anything similar.

**Comment by MarkBaker** [comment]
When you load the spreadsheet it will load the styles information "as is", with the font name, and the references to that stle for all the elements that use it (unless you tell PhpSpreadsheet not to load style information with the readDataOnly option). Nothing in the library will change that style information unless you do so explicitly.

Is it possible to send me copies of the template and the 2.xlsx files so that I can take a look?

**Comment by aqilhex** [comment]
Yes sure @MarkBaker , Here is my code as well as the excel files.

[excel.zip](https://github.com/PHPOffice/PhpSpreadsheet/files/8519999/excel.zip)

**Comment by MarkBaker** [comment]
This is what I see in your template:
![image](https://user-images.githubusercontent.com/770298/164194361-3583b924-35b9-4f77-b9b8-71db6d6893d4.png)
And this is what I see in output.xlsx
![image](https://user-images.githubusercontent.com/770298/164194670-be5099ee-0271-41fd-8842-2817bf2d0a9f.png)
If I run the code myself, and look at the output file that I'm generating, I see:
![image](https://user-images.githubusercontent.com/770298/164194949-0421365f-2c76-4d3f-8408-0fed8a96596c.png)

**Comment by aqilhex** [comment]
Try to apply any custom font, And the output will be different font face.

**Comment by MarkBaker** [comment]
While I don't have the B Nazanin font installed; I'm guessing that MS Excel is doing its best to render appropriately

**Comment by MarkBaker** [comment]
I'm running your index.php code, which adds content to cell C12, and that's also rendered as I'd expect to see it; and if I add new content manually in MS Excel, it still appears as I'd expect it to appear... so I still don't really understand what the problem is.

Can you tr posting screenshots of your before/after, to show me what difference you're seeing?

**Comment by aqilhex** [comment]
Template.xlsx
![template](https://user-images.githubusercontent.com/21981665/164198418-a4b9016a-2a4c-4338-a063-7e02c33ffac3.png)

output.xlsx
![output](https://user-images.githubusercontent.com/21981665/164198588-63aa801d-295f-4e38-931a-f3a8e8db4444.png)

**Comment by MarkBaker** [comment]
Unfortunately I don't see the difference; though the characters look "thinner": so I'm guessing that MS Excel is not finding the installed B Nazanin and falling back to its best efforts to render appropriately.

Looking at the actual xml for the font styles in the template, I see:
![image](https://user-images.githubusercontent.com/770298/164200951-2eaddd15-40a8-43c9-8ab1-973e1a148c16.png)
And for the PhpSpreadsheet generated output;
![image](https://user-images.githubusercontent.com/770298/164201116-c9831955-ca11-4024-b07d-51e532aa1d5b.png)

PhpSpreadsheet has dropped one font entry that isn't used; and it adds a few elements such as bold `b`, italic `i`, etc where Excel itself uses default values.

The only significant difference that I see there is the `charset` value. I don't know the purpose of that, so I'd need to read the OfficeOpenXML documentation to find out what it means.

**Comment by aqilhex** [comment]
It's really so important to me to have B Nazanin applied to the output.xlsx. I'm really appreciating your help and eagerly waiting for any suggestion.

**Comment by MarkBaker** [comment]
From the OpenOfficeXML documentation:

> This element specifies the character set which is supported by the parent font. This information can be used as defined in font substitution logic to locate an appropriate substitute font when this font is not available. This information is determined by querying the font when present and shall not be modified when the font is not available.

> If this element is not present, then the character set for this font shall be assumed to be the /IEC 8859-1 character set.

Attributes | Description
---|---
 characterSet (IANA Name of Character Set) | Name of the character set associated with the font. The values allowed by this attribute are defined by the names and aliases listed in the IANA registration table. If this value is specified, the value of the val attribute is ignored.

I'm not certain what this means; I'm not a font specialist, so I'll need to do some further investigation

I do know that the charset attribute isn't recognised at all in PhpSpreadsheet; so providing support for it would require code changes in the library, and to the Xlsx Reader/Writer. I can make those changes, but don't know for certain if it would resolve the problem.

**Comment by MarkBaker** [comment]
But remember that unless the B Nazanin font is also installed on the PC where the spreadsheet is being viewed, then MS Excel will always fallback to best efforts to render appropriately.

**Comment by aqilhex** [comment]
Ok so maybe it's not recognizing the font because xampp installed on other drive rather than C? because i need it on other drive to be installed. So may you please make some changes in Reader/Writer to resolve the issue?

**Comment by MarkBaker** [comment]
There's no reason why that shouldn't work. I've just installed B Nazinin myself; and am running the following code:
```php
$reader = new \PhpOffice\PhpSpreadsheet\Reader\Xlsx();
$spreadsheet = $reader->load("./template.xlsx");

$spreadsheet->getActiveSheet()->setCellValue("C12","سلام");
$spreadsheet->getActiveSheet()->getStyle('C10:E15')
    ->getFont()
    ->applyFromArray(['name' => 'B Nazanin']);

$writer = \PhpOffice\PhpSpreadsheet\IOFactory::createWriter($spreadsheet, 'Xlsx');
$writer->save('./output.xlsx');
```

And this is the output file:
![image](https://user-images.githubusercontent.com/770298/164217651-0abbd0e5-c28e-4acb-9771-79b219288eec.png)

MS Excel is telling me that the font in that cell is B Nazanin

**Comment by MarkBaker** [comment]
If I don't set the font to B Nazanin, and just use the default Nazanin rendering, then this is the cell in the output file:
![image](https://user-images.githubusercontent.com/770298/164218048-7ccd99e5-f63c-4a8c-8b6c-ce1799e7825f.png)

**Comment by aqilhex** [comment]
yes it tells me either it's B Nazanin but it's don't get applied. unless i hit enter then it will change.

**Comment by MarkBaker** [comment]
That's very weird; and I really don't have any ideas at all. It works exactly as I'd expect it to with other fonts that I've tried (e.g. Handel Gothic). I wonder if it's a quirk of rtl

**Comment by aqilhex** [comment]
Ok thanks for help i appreciate your effort. I would be grateful if you find solution for this problem. Thanks

**Comment by samangilbert** [comment]
Hi everyone,

It's been a while since your last message, but I wanted to share a problem I had with my code and how I fixed it. As our friend mentioned earlier, the issue was that when we exported an xlsx file from another one (the original) and opened it again, the fonts were not applied. Although the font name (in my case, Mitra) was displayed when selecting the cell, the actual style of the cell remained as the default Windows font.

After some investigation, I discovered that the problem was due to the fact that the original file (which was sent to me by my client from another computer) had a different version of the Mitra font. So when I opened the file on my computer, Excel couldn't find the right version of the font.

The solution was simple - I saved the file from my computer again and then uploaded it for use in my codes. Now, when I make changes to it with PhpSpreadsheet and save it, there are no problems.

It was a straightforward fix, but it took me hours to figure out the problem. I hope this helps anyone else who may face a similar issue.

**Comment by Amin1990s** [comment]
I had the same problem, after a lot of research I found out that the problem is with the charset, add the following piece of code to the file below to solve the problem.

file:
/PHPExcel/Writer/Excel2007/Style.php

Add the following code snippet to the end of the writeFont function:
```
         // charset
         if ($pFont->getName() !== null) {
             $objWriter->startElement('charset');
             $objWriter->writeAttribute('val', 0xB2);
             $objWriter->endElement();
         }
```

**Comment by Amin1990s** [comment]
I use the old version of phpexcel.
You probably need to change the charset in the path below.

PhpSpreadsheet/Writer/Xlsx/Style.php

**Comment by oleibman** [comment]
@aqilhex I am joining this conversation very late. According to @Amin1990s, what is needed is specifying a charset in writeFont when the font is B Nazanin. I have submitted PR #4501 which does just that. Can you test with that? Does it solve your problem? I unfortunately cannot tell from the discussion thus far what the problem is that you are seeing, so I need you to confirm one way or the other so I can decide whether or not to proceed with this PR.

I have some questions about your files anyhow. All the cells on Sheet1 appear to use "Nazanin" rather than "B Nazanin", but your issue seems to be with how "B Nazanin" is handled. All the entirely uninitialized cells on Sheet2 and Sheet3 use "B Nazanin", but you don't mention that they are part of your problem. I do not know how you created the template file in its current state. That would be important information if we need to investigate further.

I also draw your attention to [a different problem](https://answers.microsoft.com/en-us/msoffice/forum/all/urgent-assistance-needed-regarding-font-issue-in/2d72b127-8a2c-48b0-8ed5-5bf7178cdace). It seems related to yours, although the symptoms are not quite the same. In the other problem, the substitution font works correctly and the native "B Nazanin" on the user's system does not. This sounds like a Microsoft problem to me (and the submitter), and might also explain yours.

**Comment by Amin1990s** [comment]
@oleibman Yes, setting the charset fixed the issue for me.

Also, it seems that this solution is not limited to the B Nazanin font — it likely applies to many (if not all) Persian and Arabic fonts as well, since they often require the correct charset to render properly in Excel.

Thanks for your help and for taking the time to look into this.

**Comment by oleibman** [comment]
@Amin1990s thank you for the feedback. I will move my change forward. It is unlikely that we can maintain a complete list of Persian/Arabic fonts. However:
- Xlsx Reader will recognize and store charset if supplied. In the template, Nazanin font indicated charset 178 in the Xml, so, when we saved the copy, Nazanin was covered (without any additional action by the caller), in addition to B Nazanin (which the PR explicitly covers).
- If a font is not covered and should be, the user can call `$spreadsheet->addFontCharset('C Nazanin', 178)` before exporting the file. I will try to document this properly for the finished PR.
